### PR TITLE
Authenticate to Kusto using MI

### PR DESCRIPTION
**Pull request by dkurepa**
https://dev.azure.com/dnceng/internal/_workitems/edit/6279

### CODE REVIEW

**Review by premun — DISMISSED** [review]

**Comment by dkurepa** [comment]
> I understand this is only a silently breaking change by the fact that it requires a different set of configuration values, right?
> 
> Does an exception get thrown when these are missing?

I added checks to the `KustoClientProvider` class, the `KustoHelper` class already had then

**Review by premun — APPROVED** [review]

**Review by premun — APPROVED** [review]

**Review by premun — DISMISSED** [review]

File: src/Microsoft.DotNet.Kusto/KustoClientProvider.cs
Comment (dougbu) [inline src/Microsoft.DotNet.Kusto/KustoClientProvider.cs:75]: why is this enforced here❔ `DatabaseName` isn't used in this method. I agree it should be checked but this location is suspect. Better done in both `ExecuteKustoQueryAsync(...)` and `ExecuteStreamableKustoQuery(...)`

File: src/Microsoft.DotNet.Kusto/KustoHelpers.cs
Comment (dougbu) [inline src/Microsoft.DotNet.Kusto/KustoHelpers.cs:193]: nit: could we use `nameof(KustoOptions.KustoClusterUri)`, `nameof(KustoOptions.KustoIngestionUri)`, and similar consistently❔

File: src/Microsoft.DotNet.Kusto/KustoClientProvider.cs
Comment (dougbu) [inline src/Microsoft.DotNet.Kusto/KustoClientProvider.cs:66]: Suggest making this a `public static` helper e.g., `GetKustoConnectionStringBuilder(OptionsMonitor<KustoOptions> options)` to avoid duplication

File: test/Microsoft.DotNet.Kusto.Tests/KustoClientProviderTests.cs
Comment (dougbu) [inline test/Microsoft.DotNet.Kusto.Tests/KustoClientProviderTests.cs:42]: nit: these blocks are difficult to read. I suggest
```suggestion
        var options = new KustoOptions
        {
            KustoClusterUri = "IGNORED-CONNECTION-STRING",
            Database = "TEST-DATABASE",
        };
        using (var client = new KustoClientProvider(
            MockOptionMonitor.Create(options),
            queryProvider.Object))
```

The options never seem to change in this class; could make them a `static readonly` field.

File: src/Microsoft.DotNet.Kusto/KustoClientProvider.cs
Comment (dkurepa) [reply src/Microsoft.DotNet.Kusto/KustoClientProvider.cs:70]: I put it there to avoid code duplication, but I agree it doesn't make a lot of sense. I think you're right, better to check right before it's usage

File: src/Microsoft.DotNet.Kusto/KustoHelpers.cs
Comment (dkurepa) [reply src/Microsoft.DotNet.Kusto/KustoHelpers.cs]: my bad, fixed

**Review by riarenas — APPROVED** [review]

**Comment by dkurepa** [comment]
@AlitzelMendez please let me know if you want me to wait for the already existing dnceng-shared->helix-service PR before merging this, so we don't change two things at the same time

**Review by AlitzelMendez — APPROVED** [review]

**Comment by AlitzelMendez** [comment]
Hi @dkurepa , thanks for asking, can you hold this one day? I will take a look at the other change tomorrow :) and after that you can merge this, is that ok?

**Review by dougbu — APPROVED** [review]
lgtm

**Comment by dkurepa** [comment]
> Hi @dkurepa , thanks for asking, can you hold this one day? I will take a look at the other change tomorrow :) and after that you can merge this, is that ok?

Sure, just let me know when you think it's OK to merge!

**Comment by AlitzelMendez** [comment]
@dkurepa, I took the changes on this pull request: https://dev.azure.com/dnceng/internal/_git/dotnet-helix-service/pullrequest/39611

when that gets merge it, feel free to merge this one :)

**Pull request merged**
